---
title: "Final Figures for corral paper"
author: "Lauren Hsu"
date: "5/19/2021"
output: 
  html_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
try(source('../R/load_pkgs.R'))
try(source('../R/utils.R'))
```

### Imports and data setup
```{r, scmix_setup}
scmix_dat <- load_all_data()[1:3]
scmix_genes.use <- Reduce(intersect,lapply(scmix_dat,rownames))
scmix_dat <- lapply(scmix_dat,'[',scmix_genes.use)
for (i in 1:3){
  names(colData(scmix_dat[[i]]))[which(names(colData(scmix_dat[[i]])) == 'cell_line')] <- 'phenoid'
}
```


#### Set up scelist

```{r, eval = F}
scelist <- list(zhengmix4eq = sce_full_Zhengmix4eq(),
                zhengmix4uneq = sce_full_Zhengmix4uneq(),
                zhengmix8eq = sce_full_Zhengmix8eq(),
                scmix_10x = scmix_dat[[1]],
                scmix_celseq = scmix_dat[[2]],
                scmix_dropseq = scmix_dat[[3]])

# eliminating any 0 count genes
scelist[['zhengmix8eq']] <- scelist[['zhengmix8eq']][which(rowSums(counts(scelist[['zhengmix8eq']])) > 0),]

scelist <- lapply(scelist, logNormCounts)

# adding in HVG subsetted SCEs to the list
for(i in 1:length(scelist)){
  dat <- scelist[[i]]
  fns <- names(scelist)
  genevarmod <- modelGeneVar(dat)
  hvg_inds <- getTopHVGs(genevarmod, fdr.threshold = .05)
  scelist[[paste0(fns[i],'_HVGsub')]] <- dat[hvg_inds,]
}

matlist <- c(lapply(scelist, counts), 
             lapply(scelist, logcounts))

names(matlist) <- c(paste0(names(scelist),'.','counts'),
                    paste0(names(scelist),'.','logcounts'))

ct_list <- lapply(lapply(scelist, colData), '[', 'phenoid')

ct_list <- c(ct_list, ct_list)
names(ct_list) <- c(paste0(names(scelist),'.','counts'),
                    paste0(names(scelist),'.','logcounts'))
```

##### Standard corral

```{r, eval = F}
corral_list <- lapply(matlist, corral)
corral_embs <- lapply(corral_list, '[[', 'v')
corral_umaps <- lapply(corral_embs, uwot::umap, n_neighbors = 30)
# corral_tsnes <- lapply(corral_embs, tsne::tsne) 
#too slow; even for 3k cells, 30 pcs took > 10min
```

##### Save 

```{r, eval = F}
save(scelist, file = '../data/1_scelist.rda')
save(corral_umaps, corral_list, corral_embs, ct_list, file = '../data/1_corral_output.rda')
```

##### prcomp-equivalent results

```{r, eval = F}
tmatlist <- lapply(matlist, t)
scaled_transposed_matlist <- lapply(tmatlist, scale, center = TRUE, scale = TRUE)

pca_list <- lapply(scaled_transposed_matlist, compsvd)
pca_embs <- lapply(pca_list, '[[', 'u')
pca_umaps <- lapply(pca_embs, uwot::umap, n_neighbors = 30)
```

```{r, eval = F}
save(pca_umaps, pca_list, pca_embs, ct_list, file = '../data/1_prcomp_output.rda')
```

#### [load corral results / scelist]

```{r}
load('../data/1_corral_output.rda')
load('../data/1_scelist.rda')
```

#### [load pca results]

```{r}
load('../data/1_prcomp_output.rda')
```


## Paper figures

### Fig 1

Fig 1: Introduction to corral

#### 1b. Cell Bench mRNA mixtures 
(Adapted from section on `corral` in "Orchestrating Single Cell Analysis")

```{r}
cbdat <- CellBench::load_all_data()

# Choosing some HVGs for PCA:
sce.8qc <- logNormCounts(cbdat$mrna_mix_sortseq)
dec.8qc <- modelGeneVar(sce.8qc)
hvgs.8qc <- getTopHVGs(dec.8qc, n=1000)

sce.8qc <- sce.8qc[hvgs.8qc,]
sce.8qc$mix <- as.factor(sce.8qc$mix)

reducedDim(sce.8qc, 'PCA') <- prcomp(t(logcounts(sce.8qc)), scale. = T, center = T)$x
sce.8qc <- corral_sce(sce.8qc, subset_row=hvgs.8qc, col.w=sizeFactors(sce.8qc))

cowplot::plot_grid(
  plot_embedding_sce(sce.8qc, which_embedding = 'PCA','mix', returngg = T, plot_title = 'PCA') + guides(color = F),
  plot_embedding_sce(sce.8qc, which_embedding = 'corral','mix', returngg = T, plot_title = 'corral') + guides(color = F)
)

gridExtra::grid.arrange(
    plotPCA(sce.8qc, colour_by="mix") + ggtitle("PCA") + guides(color = F),
    plotReducedDim(sce.8qc, "corral", colour_by="mix") + ggtitle("corral") + guides(color = F),
    ncol=2
)
```

Version with no feature selection
```{r}
sce.8qcfull <- logNormCounts(cbdat$mrna_mix_sortseq)
reducedDim(sce.8qcfull, 'PCA') <- prcomp(t(logcounts(sce.8qcfull)), scale. = T)$x
sce.8qcfull <- corral_sce(sce.8qc)

sce.8qcfull$mix <- as.factor(sce.8qc$mix)

gridExtra::grid.arrange(
    plotPCA(sce.8qcfull, colour_by="mix") + ggtitle("PCA"),
    plotReducedDim(sce.8qcfull, "corral", colour_by="mix") + ggtitle("corral"),
    ncol=2
)

cowplot::plot_grid(
  plot_embedding_sce(sce.8qcfull, which_embedding = 'PCA','mix', returngg = T, plot_title = 'PCA') + guides(color = F),
  plot_embedding_sce(sce.8qcfull, which_embedding = 'corral','mix', returngg = T, plot_title = 'corral')
)
```


#### 1c. zm4 counts and logcounts embeddings

```{r}
fig1c_include <- list('zhengmix4eq.counts','zhengmix4eq.logcounts')


cowplot::plot_grid(plotlist = lapply(fig1c_include, FUN = function(nam) plot_embedding(corral_embs[[nam]],plot_title = nam, color_vec = ct_list[[nam]][['phenoid']], returngg = TRUE, dimname = 'corral')), ncol = 2)

cowplot::plot_grid(plotlist = lapply(fig1c_include, FUN = function(nam) plot_embedding(pca_embs[[nam]],plot_title = nam, color_vec = ct_list[[nam]][['phenoid']], returngg = TRUE, dimname = 'pca')), ncol = 2)

cowplot::plot_grid(plotlist = append(lapply(fig1c_include, FUN = function(nam) plot_embedding(corral_embs[[nam]],plot_title = '', color_vec = ct_list[[nam]][['phenoid']], returngg = TRUE, dimname = 'corral') + theme(legend.position = 'none')), lapply(fig1c_include, FUN = function(nam) plot_embedding(pca_embs[[nam]],plot_title = '', color_vec = ct_list[[nam]][['phenoid']], returngg = TRUE, dimname = 'pca') + theme(legend.position = 'none'))), ncol = 2)
```



#### 1d. counts and logcounts clustering ARI

Do clustering
```{r}
# 10 pcs
corral_embs_10pcs <- lapply(corral_embs, '[', i =, j = 1:10)
pca_embs_10pcs <- lapply(pca_embs, '[', i =, j = 1:10)

corral_nngraph_clusters <- lapply(corral_embs_10pcs, clusterRows, BLUSPARAM = NNGraphParam())
pca_nngraph_clusters <- lapply(pca_embs_10pcs, clusterRows, BLUSPARAM = NNGraphParam())
```

Compute ARI

```{r}
corral_ARI_10pcs <- list()
pca_ARI_10pcs <- list()

for(name in names(corral_nngraph_clusters)){
  corral_ARI_10pcs[[name]] <- pairwiseRand(as.factor(ct_list[[name]]$phenoid), 
                                       corral_nngraph_clusters[[name]], 
                                       mode = 'index')
  pca_ARI_10pcs[[name]] <- pairwiseRand(as.factor(ct_list[[name]]$phenoid), 
                                       pca_nngraph_clusters[[name]], 
                                       mode = 'index')
}
```

Consolidate data

```{r}
ARI_10pc_df_corral <- cbind(unlist(corral_ARI_10pcs),
                          rep('corral',length(corral_ARI_10pcs)),
                          data.frame(Reduce(rbind,strsplit(names(corral_nngraph_clusters), '.', fixed = T))))

rownames(ARI_10pc_df_corral) <- paste0(rownames(ARI_10pc_df_corral), 'corral')
names(ARI_10pc_df_corral) <- c('ARI','method','dataset','inp_mat')

ARI_10pc_df_pca <- cbind(unlist(pca_ARI_10pcs),
                          rep('pca',length(pca_ARI_10pcs)),
                          data.frame(Reduce(rbind,strsplit(names(pca_nngraph_clusters), '.', fixed = T))))

rownames(ARI_10pc_df_pca) <- paste0(rownames(ARI_10pc_df_pca), 'pca')
names(ARI_10pc_df_pca) <- c('ARI','method','dataset','inp_mat')

ARI_10pc_df <- rbind(ARI_10pc_df_corral, ARI_10pc_df_pca)

ARI_df <- ARI_10pc_df[!grepl('HVG', ARI_10pc_df$dataset),]
```

Final version for paper

```{r}
fig1d_datasets <- c('scmix_10x', 'scmix_celseq', 'scmix_dropseq', 
                    'zhengmix4eq', 'zhengmix4uneq', 'zhengmix8eq')

ARI_df$ARI <- round(ARI_df$ARI, digits = 2)

ggplot(ARI_df[which(ARI_df$dataset %in% fig1d_datasets),], aes(x = method, y = ARI, fill = method)) + geom_hline(yintercept=c(.25, .5, .75), linetype="dotted", col = 'gray', size = .2)+
  theme_classic() + geom_bar(stat='identity', position = 'dodge') + 
  geom_text(aes(label = ARI), size = 3, nudge_y = -.12) +
    facet_grid(rows = vars(dataset), cols = vars(inp_mat), switch = 'y') + ggtitle('ARI: NNGraph clustering on 10pcs') + 
  coord_flip() + theme(strip.text.y.left = element_text(angle = 0), axis.ticks = element_blank()) +
  scale_fill_brewer(palette="Accent") + 
  scale_x_discrete(limits = c('pca','corral'), labels = NULL) + xlab('Dataset')
```

### Fig 3

#### 3a. biplot


zm8 

```{r}
bpc <- biplot_corral(corral_list$zhengmix8eq.counts, color_vec = scelist$zhengmix8eq$phenoid, 
                text_vec = rowData(scelist$zhengmix8eq)$symbol, text_size = 3, 
              plot_title = 'Biplot: corral on Zhengmix8eq', coords = 'svd',
              xjitter = .01, yjitter = .01)

bpc
```

zm4

```{r}
biplot_corral(corral_list$zhengmix4eq.counts, color_vec = scelist$zhengmix4eq$phenoid, 
                text_vec = rowData(scelist$zhengmix4eq)$symbol, text_size = 3, 
              plot_title = 'Biplot: corral on Zhengmix4eq',
              xjitter = .01, yjitter = .01, xpc = 1)
```


#### 3b. ridge plot

```{r}
library(ggridges)
zm8mat <- counts(scelist$zhengmix8eq)
rownames(zm8mat) <- rowData(scelist$zhengmix8eq)$symbol

cowplot::plot_grid(
  ridge_plot(zm8mat, 'HLA-DRA', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev),
  ridge_plot(zm8mat, 'GNLY', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev),
  ridge_plot(zm8mat, 'LYZ', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev),
  ridge_plot(zm8mat, 'CD74', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev),
  ncol = 2)


```

version with color matching

```{r}
pf <- corral:::.generate_palette_func(ncolors = 9, pals::alphabet2())
pf2 <- function(x) return(pf(x+1)[-1])
cs_pf <- ggplot2::discrete_scale('fill','colscale',pf2)

cowplot::plot_grid(
  ridge_plot(zm8mat, 'HLA-DRA', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev) + cs_pf,
  ridge_plot(zm8mat, 'GNLY', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev) + cs_pf,
  ridge_plot(zm8mat, 'LYZ', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev) + cs_pf,
  ridge_plot(zm8mat, 'CD74', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev) + cs_pf,
  ncol = 2)


ridge_plot(zm8mat, 'TYROBP', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev) + cs_pf

ridge_plot(zm8mat, 'S100A8', scelist$zhengmix8eq$phenoid, legend = F) + theme(axis.title.y = element_blank(), axis.text.y = element_text(size = 10), axis.title.x = element_text(size = 10)) + scale_y_discrete(limits=rev) + cs_pf
```


## Fig 4

#### 4a. scmix

##### corralm 

```{r}
scmix_integ <- list()
scmix_integ$corralm_counts <- corralm(lapply(scmix_dat, counts))
scmix_integ$corralm_logcounts <- corralm(lapply(scmix_dat, logcounts))

scmix_fplots <- list()
scmix_labs <- c(scelist$scmix_10x$phenoid,
                scelist$scmix_celseq$phenoid,
                scelist$scmix_dropseq$phenoid)

gg_corralm1 <- plot_embedding(scmix_integ$corralm_counts$v, ellipse_vec = scmix_labs, 
               color_vec = rep(rownames(scmix_integ$corralm_counts$batch_sizes),
                               scmix_integ$corralm_counts$batch_sizes[,2]),
               color_title = 'batch',
               dimname = 'corralm',
               plot_title = 'corralm: counts', returngg = T)

plot_embedding(scmix_integ$corralm_counts$v, color_vec = scmix_labs, 
               ellipse_vec = rep(rownames(scmix_integ$corralm_counts$batch_sizes),
                               scmix_integ$corralm_counts$batch_sizes[,2]),
               color_title = 'Cell line',
               dimname = 'corralm',
               plot_title = 'corralm: counts')

plot(scmix_integ$corralm_counts$d)

gg_corralm2 <- plot_embedding(scmix_integ$corralm_counts$v, ellipse_vec = scmix_labs, 
               color_vec = rep(rownames(scmix_integ$corralm_counts$batch_sizes),
                               scmix_integ$corralm_counts$batch_sizes[,2]),
               color_title = 'batch',
               dimname = 'corralm',
               plot_title = '', xpc = 2, returngg = T)

dataset_vec <- rep(rownames(scmix_integ$corralm_counts$batch_sizes),
                               scmix_integ$corralm_counts$batch_sizes[,2])

scmix_fplots$corralm1 <- gg_corralm1 + theme(legend.position = "none", 
                                   axis.text = element_text(size=6), 
                                   axis.title=element_text(size=8))
scmix_fplots$corralm2 <- gg_corralm2 + theme(legend.position = "none", 
                                   axis.text = element_text(size=6), 
                                   axis.title=element_text(size=8))
```


##### SCTransform: seurat

```{r}
scmix_matlist <- lapply(scmix_dat, counts)

seurat.list <- lapply(scmix_matlist, CreateSeuratObject)

for (i in 1:length(seurat.list)) {
  seurat.list[[i]] <- SCTransform(seurat.list[[i]], verbose = TRUE)
}
  
features <- SelectIntegrationFeatures(object.list = seurat.list, nfeatures = 3000)
seurat.list <- PrepSCTIntegration(object.list = seurat.list, anchor.features = features, verbose = TRUE)

anchors <- FindIntegrationAnchors(object.list = seurat.list, normalization.method = "SCT", 
                                  anchor.features = features, verbose = TRUE)
integrated <- IntegrateData(anchorset = anchors, normalization.method = "SCT", 
                            verbose = TRUE)

integrated <- RunPCA(object = integrated)
seurat_emb <- Embeddings(integrated)[,1:10]
```

```{r}
gg_Seurat1 <- plot_embedding(seurat_emb, color_vec = dataset_vec, ellipse_vec = scmix_labs,
               plot_title = 'SCTransform + integration (Seurat)',
               dimname = 'SeuratDim',
               color_title = 'batch', xpc = 1, returngg = T)
gg_Seurat2 <- plot_embedding(seurat_emb, color_vec = dataset_vec, ellipse_vec = scmix_labs,
               plot_title = '',
               dimname = 'SeuratDim',
               color_title = 'batch', xpc = 2, returngg = T)

scmix_fplots$Seurat1 <- gg_Seurat1  + theme(legend.position = "none", 
                                   axis.text = element_text(size=6), 
                                   axis.title=element_text(size=8))
scmix_fplots$Seurat2 <- gg_Seurat2  + theme(legend.position = "none", 
                                   axis.text = element_text(size=6), 
                                   axis.title=element_text(size=8))
```


##### Harmony + corralm

```{r}
install_github("immunogenomics/harmony")
library(harmony)

dataset_vec <- rep(rownames(scmix_integ$corralm_counts$batch_sizes),
                               scmix_integ$corralm_counts$batch_sizes[,2])

scmix_integ$corralm_countsHM <- HarmonyMatrix(scmix_integ$corralm_counts$v, dataset_vec, do_pca = FALSE)
scmix_integ$corralm_logcountsHM <- HarmonyMatrix(scmix_integ$corralm_logcounts$v, dataset_vec, do_pca = FALSE)

gg_corralmHM1<- plot_embedding(scmix_integ$corralm_countsHM, color_vec = dataset_vec, ellipse_vec = scmix_labs,
               plot_title = 'corralm + HarmonyMatrix',
               dimname = 'corralm_harmony',
               color_title = 'batch', xpc = 1, returngg = T)
gg_corralmHM2<- plot_embedding(scmix_integ$corralm_countsHM, color_vec = dataset_vec, ellipse_vec = scmix_labs,
               plot_title = '',
               dimname = 'corralm_harmony',
               color_title = 'batch', xpc = 2, returngg = T)

scmix_fplots$corralmHM1 <- gg_corralmHM1 + theme(legend.position = "none", 
                                   axis.text = element_text(size=6), 
                                   axis.title=element_text(size=8))
scmix_fplots$corralmHM2 <- gg_corralmHM2 + theme(legend.position = "none", 
                                   axis.text = element_text(size=6), 
                                   axis.title=element_text(size=8))
```

##### HarmonyMatrix

```{r}
scmix_integ$HM <- HarmonyMatrix(Reduce(cbind, lapply(scmix_dat, logcounts)), dataset_vec)
scmix_integ$HMcounts<- HarmonyMatrix(Reduce(cbind, lapply(scmix_dat, counts)), dataset_vec)
gg_HM1 <- plot_embedding(scmix_integ$HM, color_vec = dataset_vec, ellipse_vec = scmix_labs,
               plot_title = 'HarmonyMatrix',
               dimname = 'HarmonyMatrix',
               color_title = 'batch', xpc = 1, returngg = T)
gg_HM2 <- plot_embedding(scmix_integ$HM, color_vec = dataset_vec, ellipse_vec = scmix_labs,
               plot_title = '',
               dimname = 'HarmonyMatrix',
               color_title = 'batch', xpc = 2, returngg = T)

scmix_fplots$HM1 <- gg_HM1 + theme(legend.position = "none", 
                                   axis.text = element_text(size=6),
                                   axis.title=element_text(size=8))
scmix_fplots$HM2 <- gg_HM2 + theme(legend.position = "none", 
                                   axis.text = element_text(size=6), 
                                   axis.title=element_text(size=8))


```


##### benchmarking

```{r}
library(rbenchmark)
HMmat <- Reduce(cbind, lapply(scmix_dat, logcounts))

scmix_time <- benchmark(corralm = {corralm(lapply(scmix_dat, counts))},
                        Seurat = {for (i in 1:length(seurat.list)) {
                          seurat.list[[i]] <- SCTransform(seurat.list[[i]], verbose = TRUE)
                        }
                          
                          features <- SelectIntegrationFeatures(object.list = seurat.list, nfeatures = 3000)
                          seurat.list <- PrepSCTIntegration(object.list = seurat.list, anchor.features = features, verbose = TRUE)
                          
                          anchors <- FindIntegrationAnchors(object.list = seurat.list, normalization.method = "SCT", 
                                                            anchor.features = features, verbose = TRUE)
                          integrated <- IntegrateData(anchorset = anchors, normalization.method = "SCT", 
                                                      verbose = TRUE)
                          
                          integrated <- RunPCA(object = integrated)},
                        Harmony = {HarmonyMatrix(HMmat, dataset_vec)})

save(scmix_time, file ='scmix_time.rda')
```


#### 4b. scal var plots

```{r}
maxY <- 1.6

svplots <- list()
svplots$corralm_counts <- scal_var(scmix_integ$corralm_counts, plot_subtitle = 'corralm on counts', max_y = maxY, returngg = TRUE)
svplots$corralm_logcounts <- scal_var(scmix_integ$corralm_logcounts, plot_subtitle = 'corralm on logcounts', max_y = maxY, returngg = TRUE)

bv <- rep(rownames(scmix_integ$corralm_counts$batch_sizes),scmix_integ$corralm_counts$batch_sizes[,2])

svplots$harmony_counts <- scal_var(scmix_integ$HMcounts, bv, plot_subtitle = 'Harmony with PCA, counts', max_y = maxY, returngg = TRUE)

svplots$harmony_logcounts <- scal_var(scmix_integ$HM, bv, plot_subtitle = 'Harmony with PCA, logcounts', max_y = maxY, returngg = TRUE)

svplots$harmonycorralm_counts <- scal_var(scmix_integ$corralm_countsHM, bv, plot_subtitle = 'Harmony with corralm, counts', max_y = maxY, returngg = TRUE)

svplots$harmonycorralm_logcounts <- scal_var(scmix_integ$corralm_logcountsHM, bv, plot_subtitle = 'Harmony with corralm, logcounts', max_y = maxY, returngg = TRUE)

svplots$sctransform <- scal_var(seurat_emb, batchvec = bv, plot_subtitle = 'SCTransform, counts', max_y = maxY, returngg = TRUE)

cowplot::plot_grid(plotlist = svplots, nrow = 1)

svplots_clean <- lapply(svplots, FUN = function(x) x + 
                          theme(legend.position = 'none', 
                                plot.title = element_blank(),
                                axis.title.y = element_blank(),
                                axis.title.x = element_blank(),
                                plot.subtitle = element_text(size=8)))

cowplot::plot_grid(plotlist = svplots_clean, nrow = 1)

#cowplot::plot_grid(plotlist = svplots_clean, ncol = 2)
```

